index.js

```diff
@@ -117,7 +123,87 @@ function adapter(uri, opts){
 
   Redis.prototype.broadcast = function(packet, opts, remote){
     Adapter.prototype.broadcast.call(this, packet, opts);
-    if (!remote) pub.publish(key, msgpack.encode([packet, opts]));
+    if (!remote) {
+      if (opts.rooms) {
+        opts.rooms.forEach(function(room) {
+          pub.publish(prefix + '#' + packet.nsp + '#' + room + '#', msgpack.encode([uid, packet, opts]));
+        });
+      } else {
+        pub.publish(prefix + '#' + packet.nsp + '#', msgpack.encode([uid, packet, opts]));
+      };
+    };
+  };
+
+  Redis.prototype.add = function(id, room, fn){
+    var self = this;
+
+    debug('adding ', id, ' to ', room);
+
+    this.sids[id] = this.sids[id] || {};
+    this.sids[id][room] = true;
+    this.rooms[room] = this.rooms[room] || {};
+    this.rooms[room][id] = true;
+
+    sub.subscribe(prefix + '#' + this.nsp.name + '#' + room + '#', function(err){
+      if (err) self.emit('error', err);
+
+      if (fn) process.nextTick(fn.bind(null, null));
+    });
+  };
+
+  Redis.prototype.del = function(id, room, fn){
+    var self = this;
+
+    debug('removing ', id, ' from ', room);
+
+    this.sids[id] = this.sids[id] || {};
+    this.rooms[room] = this.rooms[room] || {};
+    delete this.sids[id][room];
+    delete this.rooms[room][id];
+
+    if (this.rooms.hasOwnProperty(room) && !Object.keys(this.rooms[room]).length) {
+      delete this.rooms[room];
+
+      return sub.unsubscribe(prefix + '#' + this.nsp.name + '#' + room + '#', function(err){
+        if (err) self.emit('error', err);
+
+        if (fn) process.nextTick(fn.bind(null, null));
+      });
+    }
+
+    if (fn) process.nextTick(fn.bind(null, null));
+  };
+
+  Redis.prototype.delAll = function(id, fn){
+    var self = this,
+        rooms = this.sids[id];
+
+    debug('removing ', id, ' from all rooms');
+
+    if (!rooms) return process.nextTick(fn.bind(null, null));
+
+    async.forEach(Object.keys(rooms), function (room, next) {
+      if (rooms.hasOwnProperty(room)) {
+        delete self.rooms[room][id];
+      }
+
+      if (self.rooms.hasOwnProperty(room) && !Object.keys(self.rooms[room]).length) {
+        delete self.rooms[room];
+
+        return sub.unsubscribe(prefix + '#' + self.nsp.name + '#' + room + '#', function(err){
+          if (err) self.emit('error', err);
```

Comment:
If there's an error, instead of `next`ing we should abort and let the user retry the whole `delAll` if necessary (since it's idempotent)